lib/vrl/compiler/src/expression/op.rs

```diff
@@ -100,6 +101,12 @@ impl Expression for Op {
             // ... || ...
             Or => merged_def,
 
+            // {} | {}
+            Union if lhs_kind.is_object() && rhs_kind.is_object() => merged_def,
+
+            // ... | ...
+            Union  => merged_def.fallible(),
```

Comment:
With https://github.com/timberio/vector/pull/6471 merged, it would be nice to add a specific compile-time error for this case, that explains what's going on here.

Right now, it would just result in a generic "this expression can fail, resolve the failure to guarantee runtime success".